lib/instrumenter/README.md

```diff
@@ -1,3 +1,65 @@
 # Storybook Instrumenter
 
-The Storybook Instrumenter is used to patch a (3rd party) module to track and intercept function invocations for step-through debugging using the Interactions addon.
+The Storybook Instrumenter is used to patch a (3rd party) module to track and intercept function invocations for step-through debugging using the Interactions addon. In essense, the Instrumenter traverses a given object, recursively monkey-patching any functions to make them "tracked".
+
+During normal operation, tracked functions simply call through to their original function, forwarding the return value. As a side-effect, they also emit a `call` event whenever they are invoked.
+
+Through `options`, functions can be marked "interceptable", which give them another mode of operation. In this "intercept" mode, the original function is _not_ invoked, instead the interceptable function returns a `Promise` which only resolves when receiving an event to do so. This enables step-through debugging, directly in the browser. A consequence of this design is that all interceptable functions must be `await`-ed, even if their original function is not asynchronous (i.e. it normally does not return a Promise).
+
+## API
+
+The primary way to use the Storybook Instrumenter is through the `instrument` function:
+
+```ts
+instrument<TObj extends Record<string, any>>(obj: TObj, options: Options): TObj
+```
+
+`instrument` takes a plain JS object or imported ES module, and optionally an `options` object. It traverses the input object, recursively iterating over object properties and arrays. Any values with typeof `function` are tracked (through monkey-patching). Finally, a shallow copy of the original object is returned (with functions replaced). If the `mutate: true` option is set, the original object is mutated instead of returning a shallow copy.
+
+### Options
+
+- `intercept: boolean | ((method: string, path: Array<string | CallRef>) => boolean)` - Whether to make functions interceptable
+- `retain: boolean` - Whether to retain calls across renders (e.g. for story setup functions / loaders that run only once)
```

Comment:
I don't understand what this means.. can you provide a bit more details please?

Reply:
Yeah it's a tricky one. It's used solely for automatic addon-actions spies, which are instrumented though `argsEnhancers`. Those run as soon as the story is loaded into the store, regardless of whether the story is rendered, and then never again. Retain keeps those calls around in memory even when switching between stories.

Reply:
https://github.com/storybookjs/storybook/blob/e7ad10847d23c942bf99a3e8d34a06c852d9eb46/addons/interactions/src/preset/preview.ts#L12